### ISSUE TITLE
Register URL scheme / Info.plist

### ISSUE THREAD

**Issue by themihai** [murlokswarm/macpack#6]
How can I register custom URL schema? For example I'm developing a "document viewer" application so I would like  the  .doc documents and links that starts with a specific scheme such "myapp://"  (e.g. myapp://volumes/storage  )to be opened with my app and somehow handle the that input. It looks like this could be done by adding some settings in info.plist . Does macpack /murlock support a such feature? https://developer.apple.com/library/content/documentation/iPhone/Conceptual/iPhoneOSProgrammingGuide/Inter-AppCommunication/Inter-AppCommunication.html

**Comment by maxence-charriere** [comment]
yeah, in the mac.json under the field ```supported-files```: you can add the UTI corresponding to the file you want to handle.
You may also have to set the ```role``` field to ```Viewer```

Here is an example:
```json 
{
  "name": "hello",
  "id": "maxence.hello",
  "version": "1.0",
  "build-number": 1,
  "icon": "logo.png",
  "dev-region": "en",
  "deployment-target": "10.12",
  "copyright": "Copyright © 2017 maxence. All rights reserved",
  "role": "Viewer",
  "category": "",
  "sandbox": true,
  "capabilities": {
    "network": {
      "in": false,
      "out": true
    },
    "hardware": {
      "camera": false,
      "microphone": false,
      "usb": false,
      "printing": false,
      "bluetooth": false
    },
    "app-data": {
      "contacts": false,
      "location": false,
      "calendar": false
    },
    "file-access": {
      "user-selected": "",
      "downloads": "",
      "pictures": "",
      "music": "",
      "movies": ""
    }
  },
  "store": false,
  "sign-id": "",
  "supported-files": [
    "public.jpeg",
    "public.png",
    "com.compuserve.gif"
  ]
}
```

You need to launch the .app to see it in the finder when you right click.

**Comment by joeblew99** [comment]
Can i add to this as a Feature Request.

There would be a few levels for which this composition is needed:
- Between Applications that do not know about each other at compile time. 
	- Url schemes are a standard way to do this. For example: openUrl(Url("mailto:?subject=File from our app.&attach=" + fileName));
- Between components that do not know about each other in the same App.
	- Each component publishes its url schemes, and others consume them.
	- We can make this more strongly typed too
		- Go generate at compile time the golang code from the Url schemes each app publishes in their manifest.
			- Golang constants for all Components to consume at compiel time.
			- Plist and Android manifests for the whole app.
	- A third party app can send an Url scheme event, and the App can then reroute it to the appropriate Component screen.

**Comment by maxence-charriere** [comment]
I dont really understand what you trying to describe in the first point.
If you want to open a file from a url scheme, just parse the url given in ```app.OnURLOpen``` and open the file with standard Go.

For the second point, I have the feeling that it answers the question of "how component can communicates between each other?"
It can be done in multiple way (I currently use flux design pattern to do it). I would avoid to influence people to choose a specific way by having it implemented in the core of the framework. Also it could be a third party package that people choose to use or not.

For the plist, I'm not fond of it controlling what the app launches. It's a design decision I did not like back in the day I was doing Windows devs.
Plist, app manifests, there are required to describe an app and their capabilities. Describe should be the only thing they are doing.

**Comment by joeblew99** [comment]
let me try again.

Say you have 3 applications.
Each app does different things, and they have no idea about each other.
This means that an URL-scheme of app://x/y pushed out form one app can be possibly handled by any app. Sometimes more than one.

A simple example of this is how android works using intents.
If you are in an app that happens to have a http url in it and you click it, it will show you all apps that can handle that URL. In my case because i have 3 browsers installed it will show all 3.

Android intent is really just url schemes under the hood. Thats why i call this url schemes.

---

In regards to flux, and deep ULR paths into an application. yeah thats one way.
Probably easier just to handle all possible call depths in, and load up the required html view that the app determines is correct

**Comment by maxence-charriere** [comment]
So if I understand, when you click on a url scheme, you are saying that there should be something like a context menu with a list of all the app able to open this url scheme?

**Comment by joeblew99** [comment]
well you would not do it that way in the GUI, but for a demo showing off whats happening then yes.



here is a good web page that describes why and so how better than i could.
I noticed he called it service discovery for Apps and the web, which is a pretty good way to sum it up.
https://paul.kinlan.me/every-browser-should-support-intent-urls/

He also mentions Android’s Intent URLs as a good base for the abstraction. I tend to agree.
see the basic example pseudo code.

Also see the web page he referenced here:
https://paul.kinlan.me/service-discovery-on-the-web/
Its good !
